File: testcases_T2/solution.py
import sys
from collections import deque

def solve():
    """
    Solusi lengkap dan benar untuk masalah Mesin Ketik Bu Sari (T2).
    - Menggunakan deque(maxlen=20) untuk riwayat yang efisien.
    - Hanya menyimpan operasi KETIK dan HAPUS di riwayat.
    - Menyimpan karakter yang dihapus saat HAPUS untuk bisa di-undo.
    """
    try:
        # Baca jumlah operasi
        n_str = sys.stdin.readline()
        if not n_str: return
        n = int(n_str)

        # Inisialisasi buffer dan riwayat
        buffer = []
        # Deque dengan maxlen=20 secara otomatis menangani batas 20 riwayat
        # dengan membuang elemen tertua saat penuh.
        history = deque(maxlen=20) 
        MAX_BUFFER = 500

        for _ in range(n):
            line = sys.stdin.readline().strip().split()
            if not line: continue
            command = line[0]

            # --- KETIK ---
            if command == "KETIK":
                char_to_add = line[1]
                if len(buffer) < MAX_BUFFER:
                    buffer.append(char_to_add)
                    # Riwayat hanya menyimpan tuple (operasi, data_tambahan)
                    # Untuk KETIK, data tambahan tidak diperlukan.
                    history.append(("KETIK", None)) 
                    print("OK")
                else:
                    print("BUFFER PENUH")
            
            # --- HAPUS ---
            elif command == "HAPUS":
                if buffer:
                    last_char = buffer.pop()
                    # Simpan karakter yang dihapus agar bisa dikembalikan saat BATAL
                    history.append(("HAPUS", last_char))
                    print("OK")
                else:
                    print("BUFFER KOSONG")

            # --- BATAL (Undo) ---
            elif command == "BATAL":
                if history:
                    last_op, data = history.pop() # Ambil operasi terakhir dari riwayat
                    
                    if last_op == "KETIK":
                        # Membatalkan KETIK berarti menghapus karakter terakhir
                        if buffer: buffer.pop()
                    elif last_op == "HAPUS":
                        # Membatalkan HAPUS berarti mengembalikan karakter yang disimpan
                        buffer.append(data)
                    
                    print("DIBATALKAN")
                else:
                    print("TIDAK ADA OPERASI")

            # --- LIHAT ---
            # Operasi ini tidak mengubah state dan tidak masuk riwayat
            elif command == "LIHAT":
                if buffer:
                    print("".join(buffer))
                else:
                    print("KOSONG")

            # --- CARI ---
            # Operasi ini tidak mengubah state dan tidak masuk riwayat
            elif command == "CARI":
                word_to_find = line[1]
                current_buffer_str = "".join(buffer)
                pos = current_buffer_str.find(word_to_find)
                if pos != -1:
                    print(f"DITEMUKAN {pos}")
                else:
                    print("TIDAK DITEMUKAN")
    
    except (IOError, ValueError, IndexError):
        # Penanganan error jika input tidak sesuai/kosong
        return

if __name__ == "__main__":
    solve()

File: testcases_T2/solution.cpp
#include <iostream>
#include <string>
#include <vector>
#include <deque>

// Struct untuk menyimpan data riwayat secara jelas
// Ini kunci untuk membatalkan operasi HAPUS dengan benar
struct HistoryEntry {
    std::string operation; // "KETIK" atau "HAPUS"
    char character;      // Menyimpan karakter yang dihapus. Tidak digunakan untuk KETIK.
};

// Fungsi untuk mempercepat I/O di C++
void setup_io() {
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(NULL);
}

int main() {
    setup_io();

    int n;
    std::cin >> n;

    std::string buffer = "";
    // Menggunakan deque agar efisien menghapus elemen tertua (pop_front)
    std::deque<HistoryEntry> history;
    const int MAX_HISTORY = 20;
    const int MAX_BUFFER = 500;

    std::string command;
    // Loop untuk setiap operasi
    while (n-- > 0 && std::cin >> command) {
        
        // --- KETIK ---
        if (command == "KETIK") {
            char c;
            std::cin >> c;
            if (buffer.length() < MAX_BUFFER) {
                buffer += c;
                // Hanya KETIK yang masuk riwayat
                if (history.size() == MAX_HISTORY) {
                    history.pop_front(); // Buang riwayat tertua jika penuh
                }
                history.push_back({"KETIK", '\0'}); // Char tidak penting untuk KETIK
                std::cout << "OK\n";
            } else {
                std::cout << "BUFFER PENUH\n";
            }
        
        // --- HAPUS ---
        } else if (command == "HAPUS") {
            if (!buffer.empty()) {
                char last_char = buffer.back();
                buffer.pop_back();
                // Hanya HAPUS yang masuk riwayat, simpan juga karakternya
                if (history.size() == MAX_HISTORY) {
                    history.pop_front();
                }
                history.push_back({"HAPUS", last_char});
                std::cout << "OK\n";
            } else {
                std::cout << "BUFFER KOSONG\n";
            }
        
        // --- BATAL ---
        } else if (command == "BATAL") {
            if (!history.empty()) {
                HistoryEntry last_op = history.back();
                history.pop_back();
                
                if (last_op.operation == "KETIK") {
                    // Membatalkan KETIK -> hapus karakter
                    if (!buffer.empty()) {
                        buffer.pop_back();
                    }
                } else if (last_op.operation == "HAPUS") {
                    // Membatalkan HAPUS -> kembalikan karakter yang disimpan
                    buffer += last_op.character;
                }
                std::cout << "DIBATALKAN\n";
            } else {
                std::cout << "TIDAK ADA OPERASI\n";
            }
        
        // --- LIHAT ---
        // Tidak masuk riwayat
        } else if (command == "LIHAT") {
            if (buffer.empty()) {
                std::cout << "KOSONG\n";
            } else {
                std::cout << buffer << "\n";
            }
        
        // --- CARI ---
        // Tidak masuk riwayat
        } else if (command == "CARI") {
            std::string word;
            std::cin >> word;
            size_t pos = buffer.find(word); // size_t adalah tipe data yang tepat
            if (pos != std::string::npos) {
                std::cout << "DITEMUKAN " << pos << "\n";
            } else {
                std::cout << "TIDAK DITEMUKAN\n";
            }
        }
    }

    return 0;
}